src/main.rs

```diff
@@ -170,16 +172,24 @@ fn sync(lockfile: &Path,
         // Next up, copy it to the vendor directory
         let name = format!("{}-{}", id.name(), id.version());
         let src = src.join(&name).into_path_unlocked();
-        let dst_name = if !explicit_version && id.version() == max[id.name()] {
-            id.name().to_string()
-        } else {
+        let dir_has_version_suffix = explicit_version || id.version() != max[id.name()];
+        let dst_name = if dir_has_version_suffix {
+            // Eg vendor/futures-0.1.13
             format!("{}-{}", id.name(), id.version())
+        } else {
+            // Eg vendor/futures
+            id.name().to_string()
         };
         let dst = local_dst.join(&dst_name);
+
         let cksum = dst.join(".cargo-checksum.json");
-        if cksum.exists() {
+        if dir_has_version_suffix && cksum.exists() {
+            // Always re-copy directory without version suffix in case the version changed
```

Comment:
Should we open up the Cargo.toml to check if the version has actually changed? That would prevent a lot of FS copying, but I'm not sure it matters too much. Vendoring is not typically on the path that needs to be fast.

In my case, I wouldn't mind unconditionally re-vendoring everything (eg wiping the vendor/ directory) every time. But this at least fixes the bug.